spec/src/main/asciidoc/interoperability.asciidoc

```diff
@@ -35,34 +35,45 @@ To meet those requirements, we introduce 2 new claims to the MP-JWT:
 exp:: Identifies the expiration time on or after which the JWT MUST NOT be accepted for processing.  The processing of the
     "exp" claim requires that the current date/time MUST be before the expiration date/time listed in the "exp" claim.
     Implementers MAY provide for some small leeway, usually no more than a few minutes, to account for clock skew.
     Its value MUST be a number containing a NumericDate value. https://tools.ietf.org/html/rfc7519#section-4.1.4[RFC7519, Section 4.1.4]
-iat:: Identifies the time at which the JWT was issued.  This claim can be used to determine the age of the JWT.  Its
-    value MUST be a number containing a NumericDate value. https://tools.ietf.org/html/rfc7519#section-4.1.6[RFC7519, Section 4.1.6]
+upn (or preferred_username or sub):: MP-JWT "upn" claim is the user principal name in the java.security.Principal interface, and is the caller
+    principal name in javax.security.enterprise.identitystore.IdentityStore. If this claim is missing, fallback to
+    the http://openid.net/specs/openid-connect-core-1_0.html#StandardClaims["preferred_username", OIDC Section 5.1] should be
+    attempted, and if that claim is missing, fallback to the "sub" claim should be used.
+
+#### Recommended MP-JWT claims
+
+sub:: Uniquely identifies the principal that is the subject of the JWT. See the "upn" claim for how this relates to the container
+ java.security.Principal. https://tools.ietf.org/html/rfc7519#section-4.1.2[RFC7519, Section 4.1.2].
 jti:: Provides a unique identifier for the JWT. The identifier value MUST be assigned in a manner that ensures that
     there is a negligible probability that the same value will be accidentally assigned to a different data object;
     if the application uses multiple issuers, collisions MUST be prevented among values produced by different issuers
     as well.  The "jti" claim can be used to prevent the JWT from being replayed.  The "jti" value is a case-sensitive
     string. https://tools.ietf.org/html/rfc7519#section-4.1.7[RFC7519, Section 4.1.7]
-upn:: This MP-JWT custom claim is the user principal name in the java.security.Principal interface, and is the caller
-    principal name in javax.security.enterprise.identitystore.IdentityStore. If this claim is missing, fallback to
-    the http://openid.net/specs/openid-connect-core-1_0.html#StandardClaims["preferred_username", OIDC Section 5.1] should be
-    attempted, and if that claim is missing, fallback to the "sub" claim should be used.
+aud:: Identifies the MP JWT endpoint(s) which can be accessed by JWT. https://tools.ietf.org/html/rfc7519#section-4.1.3[RFC7519, Section 4.1.3]
+
+[NOTE] MP JWT implementations may enforce that JWT tokens contain all the recommended headers and claims. The recommended headers and claims may become required in the future versions of the MP JWT specification.
```

Comment:
Does this mean that a MP JWT implementation may force for all recommended claims to be mandatory and another  to have them as optionals?

Reply:
@radcortez yes since `recommended` is `optional`.  The message here is that if a given implementation has an idea what to do about `jti` and the replay protection for example then it is encouraged to do it. Not sure, may be the language can become clearer :-)

Reply:
No, I think the text is clear, but may concern is that the JWT may not be portable across MP JWT impls. I'm thinking in scenarios where you have a JWT that is accepted on a service, propagates it to another service with a MP JWT implementation with stricter rules and then fail the JWT. I guess one way to fix it is to make the 2nd service happy, which should also work on the 1st.

On the other hand, since this is somehow open, you have have mix behaviour where some claims / headers are mandatory in an implementation and the other way around in another.

I'm not completely sure if we should leave this to the implementation. I see some potential issues here. What do you think?

Reply:
This is like everything else where the spec lets the implementations do something optionally (example - some signature algorithms), if we leave the text out then it will be still the case with for example with the implementation X being relaxed about say the token age limit, the same token will be blocked by the downstream smallrye-jwt endpoint taking a harder stance :-).
The deployments involving potentially multiple vendor MP JWT impls would not be dynamic - the policy will have to be aligned. Which is what you suggested :-).

Reply:
If everyone else is ok with this, I'm also fine. Just wanted to raise the flag and clarify :)